src/Plugin/views/query/CivicrmSql.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+namespace Drupal\civicrm_entity\Plugin\views\query;
+
+use Drupal\civicrm\Civicrm;
+use Drupal\Core\Database\Database;
+use Drupal\views\Plugin\views\display\DisplayPluginBase;
+use Drupal\views\Plugin\views\query\Sql;
+use Drupal\views\ViewExecutable;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Views query plugin for a CiviCRM Entity SQL query.
+ *
+ * @ingroup views_query_plugins
+ *
+ * @ViewsQuery(
+ *   id = "civicrm_views_query",
+ *   title = @Translation("CiviCRM SQL Query"),
+ *   help = @Translation("Query will be generated and run using the Drupal database API against the CiviCRM database.")
+ * )
+ */
+class CivicrmSql extends Sql {
+
+  /**
+   * The CiviCRM service.
+   *
+   * @var \Drupal\civicrm\Civicrm
+   */
+  protected $civicrm;
+
+  /**
+   * Set the CiviCRM service.
+   *
+   * @param \Drupal\civicrm\Civicrm $civicrm
+   *   The CiviCRM service.
+   *
+   * @note we use this pattern to avoid constructor overrides.
+   */
+  public function setCivicrm(Civicrm $civicrm) {
+    $this->civicrm = $civicrm;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    $instance = parent::create($container, $configuration, $plugin_id, $plugin_definition);
+    $instance->setCivicrm($container->get('civicrm'));
+    return $instance;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function init(ViewExecutable $view, DisplayPluginBase $display, array &$options = NULL) {
+    // Ensure that Drupal is aware of the CiviCRM database connection.
+    // This should be added into the settings.php, but we provide a backwards
+    // compatibility layer here.
+    // @todo can we get this upstream in the CiviCRM module on initialize?
+    $this->civicrm->initialize();
+    $connection_name = drupal_valid_test_ua() ? 'civicrm_test' : 'civicrm';
+    if (!Database::getConnectionInfo($connection_name)) {
+      $civicrm_connection_info = Database::convertDbUrlToConnectionInfo(CIVICRM_DSN, DRUPAL_ROOT);
+      Database::addConnectionInfo($connection_name, 'default', $civicrm_connection_info);
+    }
+    parent::init($view, $display, $options);
```

Comment:
This is the magic trick! It ensures the connection is available if it hasn't been defined in `settings.php`

Reply:
This is great. So much magic in one PR :)